pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go

```diff
@@ -67,12 +68,21 @@ func (c CouchDBEDVProvider) OpenStore(name string) (edvprovider.EDVStore, error)
 // CouchDBEDVStore represents a CouchDB store with functionality needed for EDV data storage.
 // It wraps an edge-core CouchDB store with additional functionality that's needed for EDV operations.
 type CouchDBEDVStore struct {
+	mux       sync.RWMutex
 	coreStore storage.Store
 }
 
 // Put stores the given document.
 // A mapping document is also created and stored in order to allow for encrypted indices to work.
-func (c CouchDBEDVStore) Put(document models.EncryptedDocument) error {
+func (c *CouchDBEDVStore) Put(document models.EncryptedDocument) error {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	err := c.checkNewDocumentForConflictingIndexWithAnyExistingDocs(document)
```

Comment:
This check isn't atomic, i.e. by the time the check finishes and no error returns another document with the same index could have been stored. What happens in this case, i.e. when you do a subsequent query using this index?

Reply:
Sorry, I see that you actually do a write lock. I don't think it's a good idea to lock the entire store for a write of a single document. The store is supposed to handle many concurrent requests (if I'm not mistaken).

Reply:
Per our discussion, I removed the lock and made a note regarding the impossibility of ensuring uniqueness due to the eventual consistency model of CouchDB. I will have to deal with the possibility of duplicate VCs in edge-service still - but I don't believe I can do anything about it here. This is just a best-effort attempt.